### PR TITLE
Feat: Expand User type

### PR DESCRIPTION
**Pull request by Bouncey**
This PR contains commits from https://github.com/freeCodeCamp/open-api/pull/48, by mistake.

This PR expands the `User` type, adjusts the user query's to return a single user and adds a challengeMap Mutation type.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,8 +21,10 @@
     "deploy-prod": "serverless --stage=prod deploy",
     "format": "prettier --write es5 './**/*.js'",
     "prepare-production": "snyk protect",
-    "start": "serverless offline start --skipCacheInvalidation",
-    "test": "snyk test"
+    "start": "DEBUG=fcc:* serverless offline start --skipCacheInvalidation",
+    "test": "jest",
```
Comment (Bouncey) [inline package.json]: Having travis run `snyk test` was breaking the build due to travis not having the required auth

Comment (ojongerius) [reply package.json]: Ah that's bad DX.

The issue is that your PR comes from your fork. Your fork does not have the secret variable used to authenticate to Snyk. Hence the build fails.

One way to handle this is to disable Snyk being run for PR builds, by testing the value of $TRAVIS_PULL_REQUEST. The upshot is that contributors can enjoy testing and green builds. The downside is that the merge build will then fail, putting load on the maintainers to silence Snyk.

Thoughts?

**Comment by ojongerius** [comment]
I'll need a little more time to review this, I can do that on my Monday, your Sunday.

File: package.json (hunk 1 above)
Comment (Bouncey) [reply package.json]: What is `snyk test` doing for us on travis that the `snyk` check isn't doing for us when a PR is opened.

The majority of PR's hopefully will come from forks as not everyone has the required permissions to push to the main repo.

Even PR's from ourselves should be coming from forks. This hopefully reduces the 'I made a bad push' problems that are possible when pushing to the main repo. Screwing up a fork is nothing compared to breaking the main repository through a few sloppy git commands.

**TLDR;** Across the organisation, we try to discourage writing directly to the main repo. Forks are friends, not ... food? I know what I mean.

File: .travis.yml [hunk 2]
```diff
@@ -12,9 +12,7 @@ env:
     - AWS_REGION=us-east-1
     - SLS_DEBUG=true
 
-before_install:
-  - npm install -g serverless
```
Comment (ojongerius) [inline .travis.yml]: AFAICT if you remove this, our deploy script will not be able to call `sls`, *unless* you update the script to point to `./node_modules/.bin/sls`

File: dataLayer/mongo/user.js [hunk 3]
```diff
@@ -43,3 +31,27 @@ export async function createUser(email) {
...
+  }
+  const { challengeMap } = user;
+  const updated = [...challengeMap, ...challengeMapUpdate];
+  try {
+    await user.update({ challengeMap: updated });
+    log('user updated');
+    return updated;
+  } catch (err) {
+    log('crashing hard');
+    // TODO(Bouncey): Communicate the problem with the client
```
Comment (ojongerius) [inline dataLayer/mongo/user.js]: I'd prefer raising an issue and add the issue number. It will be easier for the team to see what work is outstanding, and get some context on the issue at hand. 
I've seen comments in the freeCodeCamp/freeCodeCamp codebase where the person writing something like "can be removed later on" was himself unclear on under what conditions it could be removed.

File: dataLayer/mongo/user.js [hunk 4]
```diff
@@ -43,3 +31,27 @@ export async function createUser(email) {
   );
   return newUser;
 }
+
+export async function getUser(email) {
+  const user = await findUserBy({ email });
+  return applyMigrations(user);
```
Comment (ojongerius) [inline dataLayer/mongo/user.js]: Mutating state on a simple request to get a user? Are we not better off by doing migrations by batches, and asserting new users are created correctly?

File: graphql/typeDefs/User/query.js [hunk 5]
```diff
@@ -1,9 +1,6 @@
+// TODO(Bouncey): requires auth protection
 export default `
 type Query {
-  users(
-      _id: ID
-      name: String
-      email: String
-  ): [User]
+  getSessionUser(email: String!): User
```
Comment (ojongerius) [inline graphql/typeDefs/User/query.js]: I'm assuming you created this to get the user object for the current user provided the session is authenticated? How about letting user() returning the currently logged in user, and returning an empty result if here is no user authenticated?

The following query:

```graphql
query {
  user {
    id
  }
}
```

Would then return, when authenticated:

```graphql
{
  "data": {
    "user": {
      "id": "e82403ce-4687-11e8-bca3-14109fd1f8cb" 
    }
  }
}
```

Or, when not authenticated, an empty user, with a helpful error object:

```graphql
{
  "data": {
      "user": null
    },
   "errors": {
        "message": "Not logged in",
        "locations": [
            {
                "line": 2,
                "column": 2
        ],
        "path": [
            "user"
        ]
    }
}
```

File: package.json (hunk 1 above)
Comment (ojongerius) [reply package.json]: > What is snyk test doing for us on travis that the snyk check isn't doing for us when a PR is opened.

You are completely right. Nothing.

> Even PR's from ourselves should be coming from forks. This hopefully reduces the 'I made a bad push' problems that are possible when pushing to the main repo. Screwing up a fork is nothing compared to breaking the main repository through a few sloppy git commands.

This is a good point, it's so convenient to push directly, but you are right, it's better to use forks.

You've convinced me, let's get rid of it and call jest instead 🙂

File: graphql/resolvers/user.js [hunk 6]
```diff
@@ -1,10 +1,24 @@
...
+  createUser,
+  updateChallengeMap
+} from '../../dataLayer/mongo/user';
+
+const log = debug('fcc:resolvers:user');
 
 export const userResolvers = {
   Query: {
-    users: (_, args) => dbUsers.getUsers(args)
+    // TODO(Bouncey): requires auth protection
```
Comment (ojongerius) [inline graphql/resolvers/user.js]: See comments about TODO earlier on. You could probably convince me this is ok for now, as things are in a state of flux while we get this project production ready. However, you would be giving a good example for new contributors, and it will be easier for the whole team to see what's going on by using issues for this.

Comment (Bouncey) [reply graphql/resolvers/user.js]: I agree that it would be really good to have this in place now. If auth had been implemented at the project level, it would definitely be implemented for this query and the user related mutations.

In my opinion it should be a separate PR that implements some form of auth handling, and that there is no place for it in this PR. I am looking in to this on a separate branch.

**Comment by Bouncey** [comment]
@ojongerius Looking back at this, I think I could make the queries and mutations slicker by using context and not hitting the db on every request to find a user. I am going to re-write some parts of this PR. 

It may be that we need auth in place before this PR can/should progress.

File: dataLayer/mongo/user.js (hunk 4 above)
Comment (Bouncey) [reply dataLayer/mongo/user.js]: I will remove this.

For now, we should concentrate on getting an MVP assuming all users will be new, which will be the case during the QA phase. Once the platform has stabilised some-what, we can plan for existing user migrations.

File: dataLayer/mongo/user.js (hunk 3 above)
Comment (Bouncey) [reply dataLayer/mongo/user.js]: The 'issue' there is, the issue will be raised against code that potentially doesn't exist yet, which will lead to confusion. I agree that linking to as issue would be better though.


@raisedadead, is there scope for a github bot to raise an issue for each `TODO` pushed to staging? Perhaps with a link to the file and line containing the `TODO`?

**Comment by Bouncey** [comment]
Waiting on #58

**Comment by ojongerius** [comment]
#58 went in (nice work!), what's the status of this one at the moment?

**Comment by Bouncey** [comment]
Almost a rewrite, but it is nearly there. 

One issue I am having that ibmeant to ping you for is, if I use mongo in a different test suit in the same way that the users integration test does, I get a race condition between the the `afterAll` disconnects. If onebsuit is still running then the tests fail because ` MongoError: Topology was destroyed` or something to that effect.

I have tried managing a connection in the global setup/teardown, but the tests would timeout. Any ideas?

File: dataLayer/mongo/user.js (hunk 3 above)
Comment (raisedadead) [reply dataLayer/mongo/user.js]: Yes, it should be possible. I'll create a user story for this on the github bot.

**Comment by ojongerius** [comment]
@Bouncey interesting! Does this happen when you run jest with `--runInBand` too?

**Comment by Bouncey** [comment]
@ojongerius yes it does, I think it is due to the async nature of the setup and teardown, where at times, they overlap. I will push what I have so far so you can see.

I think I'm in a place where I shouldn't be adding any further commits to this PR, else the scope will slip too far.

**Comment by Bouncey** [comment]
@ojongerius I might have fixed the race condition be returning the promise in `teardown`?

File: src/dataLayer/mongo/user.js [hunk 7]
```diff
@@ -1,46 +1,67 @@
...
-  const exists = !!entities.length;
-  if (exists) {
-    throw new Error(`An account with the email ${email} already exists`);
+  const newUser = { name, email };
+  const accountLinkId = decoded[namespace + 'accountLinkId'];
+  if (accountLinkId) {
+    newUser.accountLinkId = accountLinkId;
+  } else {
+    newUser.accountLinkId = uuid();
+    // TODO: appy accountLinkId to app_metadata (auth0)
```
Comment (Bouncey) [inline src/dataLayer/mongo/user.js]: I will fix this typo by adding the code required to talk to auth0

**Comment by Bouncey** [comment]
I worte all of those test before rebasing, so glad that I did because I made a unholy mess and lost two days with of work with a clumsy `git stash pop`.

 ## Yay for tests! :tada:

**Comment by ojongerius** [comment]
Sounds good! Sounds like it's review time? Are you keeping the WiP label while finish writing the code to talk to auth0?

**Comment by Bouncey** [comment]
Please review. None of this code should change, all I need to add is an API call to auth0

**Comment by ojongerius** [comment]
👍 will have a look today.

File: package.json [hunk 8]
```diff
@@ -27,7 +27,7 @@
     "snyk-test": "snyk test",
     "start":
       "cross-env DEBUG=fcc:* nodemon node_modules/serverless/bin/serverless offline start --skipCacheInvalidation",
-    "test": "cross-env JWT_CERT=test jest --verbose",
+    "test": "cross-env JWT_CERT=test jest --runInBand  --verbose --silent",
```
Comment (ojongerius) [inline package.json:30]: I thought we agreed to have both `test` and `test-ci` run with `--verbose`?

File: src/dataLayer/model/user.js [hunk 9]
```diff
@@ -1,10 +1,13 @@
 const mongoose = require('mongoose');
-const validator = require('validator');
 
 const Schema = mongoose.Schema;
-const SchemaTypes = Schema.Types;
 
 const userSchema = new Schema({
+  accountLinkId: {
+    type: 'string',
+    description: 'A uuid used to link SSO and freeCodeCamp accounts together',
```
Comment (ojongerius) [inline src/dataLayer/model/user.js]: Is the idea here to have one globally unique UUID that we can use across services? I think that is the way forward as components get refactored into different services, and databases 👍 

If this would the case, this ID should be decoupled from both Mongo and SSO, and a name that is more neutral, like accountId for instance, would be more fitting in my opinion.

File: src/dataLayer/model/user.js [hunk 10]
```diff
@@ -253,16 +186,7 @@ const userSchema = new Schema({
...
-      coreTeam: {
-        type: 'array',
-        default: []
-      }
-    },
-    default: {}
   }
 });
 
-module.exports = mongoose.model('User', userSchema, 'user');
```
Comment (ojongerius) [inline src/dataLayer/model/user.js]: 😕 I'm surprised this works, I added 'user' here, because Mongoose would create a collection "users", unless I instructed it to use the singular "user". Just curious: why did this work?

File: src/dataLayer/model/user.js [hunk 11]
```diff
@@ -36,38 +36,16 @@ const userSchema = new Schema({
   isCheater: {
     type: 'boolean',
     description:
-      'Users who are confirmed to have broken academic honesty policy are marked as cheaters',
+      'Users who are confirmed to have broken academic honesty policy are ' +
+      'marked as cheaters',
     default: false
   },
-  isGithubCool: {
```
Comment (ojongerius) [inline src/dataLayer/model/user.js]: Cleanup!

**Review by ojongerius — COMMENTED** [review]
Started, but have not yet finished reviewing, will give this some more attention in a bit.

File: package.json (hunk 8 above)
Comment (Bouncey) [reply package.json:30]: They are? The `silent` flag squashes any console logs triggered by the tests. Unexpected errors caught by `asyncErrorHandler` are logged to the console before we throw a generic error, for example.

File: src/dataLayer/model/user.js (hunk 10 above)
Comment (Bouncey) [reply src/dataLayer/model/user.js]: Interesting, it works because we never reference the model directly. I cannot even remember why I changed this.

File: src/dataLayer/mongo/user.js [hunk 12]
```diff
@@ -1,46 +1,67 @@
...
-        resolve(res.users);
-      })
-      .catch(err => {
-        reject(err);
-      });
+export function getUser(root, { email }) {
+  return new Promise(async function getUserPromise(resolve, reject) {
+    if (!isString(email) || !validator.isEmail(email)) {
+      reject(
+        new TypeError(`Expected a valid email, got ${JSON.stringify(email)}`)
```
Comment (ojongerius) [inline src/dataLayer/mongo/user.js:20]: What does `JSON.stringify()` add here?

File: src/dataLayer/model/user.js (hunk 9 above)
Comment (ojongerius) [reply src/dataLayer/model/user.js]: Sorry I had not finished looking at all the code when I posted the comment and question above.

If I understand the code in `src/dataLayer/mongo/user.js` right this will contain a unique identifier for each user, at auth0, is that correct?

File: src/graphql/typeDefs/User/type.js [hunk 13]
```diff
@@ -1,7 +1,92 @@
...
+  path: String
+  history: [String]!,
+  name: String!
+  ext: String!
+  contents: String!
+  source: String
+}
+
+type CompletedChallenge {
+  id: String
```
Comment (ojongerius) [inline src/graphql/typeDefs/User/type.js]: Should all or some of these fields by mandatory?

File: src/graphql/typeDefs/User/type.js [hunk 14]
```diff
@@ -1,7 +1,92 @@
...
+  completedChallenge: String
+}
+
+type Streak {
+  longest: Int!
+  current: Int!
+}
+
 type User {
-  _id: ID @isAuthenticatedOnField
```
Comment (ojongerius) [inline src/graphql/typeDefs/User/type.js]: Good riddance

File: test/integration/__snapshots__/userFlow.test.js.snap [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+Object {
+  "email": "charlie@thebear.me",
+  "name": "Charlie",
+}
+`;
+
+exports[`createUser should not create a second user for the same context: no duplicate users 1`] = `
+Array [
+  Object {
+    "accountLinkId": "a-very-unique-string-for-charlie@thebear.me",
```
Comment (ojongerius) [inline test/integration/__snapshots__/userFlow.test.js.snap]: This is entertaining, but if they are UUIDs, should we use something that at least looks like a UUID?

File: test/integration/userFlow.test.js [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+      .then(({ data, errors }) => {
+        expect(data.createUser).toMatchSnapshot('new user');
+        expect(errors).toMatchSnapshot('no errors');
+        return;
+      })
+      .then(done)
+      .catch(done);
+  });
+
+  it('should not create a second user for the same context', done => {
```
Comment (ojongerius) [inline test/integration/userFlow.test.js]: I think we are testing that we are not allowed to create duplicate users, right? Should we use that in the description instead?

File: test/integration/userFlow.test.js [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+      .then(done)
+      .catch(done);
+  });
+
+  it('should return a user after one has been created', done => {
+    expect.assertions(2);
+
+    graphql(graphqlSchema, expectedUserQuery, rootValue, validContextCharlie)
+      .then(({ data, errors }) => {
+        expect(data).toMatchSnapshot('user found');
```
Comment (ojongerius) [inline test/integration/userFlow.test.js:152]: I find this much harder to read than a simple test that creates an object, uses that to create a user and validates by compares the result with the input.

Can you explain to me what added value of snapshots in this case are? I see people online rave about snapshot, and I get that it's awesome for front end testing. But what does it bring us here but harder to read code?

If it sounds like I'm nagging, I'm sorry.  I'm genuinely interested, and am probably missing the point! 😄

File: test/utils/test-environment.js [hunk 18]
```diff
@@ -13,19 +16,45 @@ class MongoEnvironment extends NodeEnvironment {
     this.global.__MONGO_URI__ = await global.__MONGOD__.getConnectionString();
     this.global.__MONGO_DB_NAME__ = global.__MONGO_DB_NAME__;
 
-    const token = jwt.sign({ id: 123, name: 'Charlie' }, JWT_CERT);
+    const token = jwt.sign(
+      {
+        id: 123,
```
Comment (ojongerius) [inline test/utils/test-environment.js]: Should we drop the id if we are not using?

File: test/utils/test-environment.js [hunk 19]
```diff
@@ -13,19 +16,45 @@ class MongoEnvironment extends NodeEnvironment {
...
+        name: 'Charlie',
+        email: 'charlie@thebear.me',
+        [namespace +
+        'accountLinkId']: 'a-very-unique-string-for-charlie@thebear.me'
```
Comment (ojongerius) [inline test/utils/test-environment.js]: change to UUID if we use those?

File: src/graphql/typeDefs/GDPR-User/mutation.js [hunk 20]
```diff
@@ -0,0 +1,5 @@
+export default `
+type Mutation {
+  deleteGDPRUser(accountLinkId: String!): HTTPStatus! @isAuthenticatedOnQuery
```
Comment (ojongerius) [inline src/graphql/typeDefs/GDPR-User/mutation.js:5]: Sorry I do not have much context on this, but would this be used by users that are logged in to (requests scheduling of) deletion of their own account?

If this is the case, should we allow calling `deleteGDPRUser` but instead of using accountLinkedId from input use the ID we get from the users context?

**Comment by ojongerius** [comment]
Finished another pass 😅 This is turned into quite an epic. Awesome work but please don't add more complexity! 😄 

FWIW tests are passing but are very noisy for me:

```sh
▶ JWT_CERT=test jest --runInBand --verbose
 PASS  ../test/integration/userFlow.test.js
  createUser
    ✓ should return null and an auth error without a token (72ms)
    ✓ should return null and an auth error with an invalid token (3ms)
    ✓ should create a user by query (73ms)
    ✓ should not create a second user for the same context (36ms)
  getUser
    ✓ should return null and an auth error without a token (3ms)
    ✓ should return null and an auth error with an invalid token (1ms)
    ✓ should return a user after one has been created (7ms)
    ✓ should return null if no user has been found (8ms)
    ✓ should return errors for a malformed query (4ms)

  console.error dataLayer/mongo/user.js:60
    Cannot create account, the unique id associated with this user is already in use a-very-unique-string-for-charlie@thebear.me

  console.log ../node_modules/graphql-tools/dist/schemaGenerator.js:514
    { Error: No user found for nowhere@tobe.seen
        at getUserPromise (/Users/ojongerius/repos/fcc-open-api/src/dataLayer/mongo/user.js:28:14)
        at <anonymous>
      originalMessage: 'No user found for nowhere@tobe.seen',
      message: 'Error in resolver Query.getUser\nNo user found for nowhere@tobe.seen' }

  console.error utils/asyncErrorHandler.js:6
    Error: No user found for nowhere@tobe.seen
        at getUserPromise (/Users/ojongerius/repos/fcc-open-api/src/dataLayer/mongo/user.js:28:14)
        at <anonymous>

  console.log ../node_modules/graphql-tools/dist/schemaGenerator.js:514
    { TypeError: Expected a valid email, got "Ooops!"
        at getUserPromise (/Users/ojongerius/repos/fcc-open-api/src/dataLayer/mongo/user.js:20:9)
        at new Promise (<anonymous>)
        at getUser (/Users/ojongerius/repos/fcc-open-api/src/dataLayer/mongo/user.js:17:10)
        at Object.<anonymous> (/Users/ojongerius/repos/fcc-open-api/node_modules/graphql-tools/src/schemaGenerator.ts:683:22)
        at class_1.<anonymous> (/Users/ojongerius/repos/fcc-open-api/node_modules/graphql-tools/src/schemaGenerator.ts:781:42)
        at step (/Users/ojongerius/repos/fcc-open-api/node_modules/graphql-tools/dist/schemaGenerator.js:51:23)
        at Object.next (/Users/ojongerius/repos/fcc-open-api/node_modules/graphql-tools/dist/schemaGenerator.js:32:53)
        at /Users/ojongerius/repos/fcc-open-api/node_modules/graphql-tools/dist/schemaGenerator.js:26:71
        at new Promise (<anonymous>)
        at Object.<anonymous>.__awaiter (/Users/ojongerius/repos/fcc-open-api/node_modules/graphql-tools/dist/schemaGenerator.js:22:12)
        at /Users/ojongerius/repos/fcc-open-api/node_modules/graphql-tools/src/schemaGenerator.ts:781:13
        at isAuthenticatedOnQuery (/Users/ojongerius/repos/fcc-open-api/src/graphql/resolvers/directives.js:24:32)
        at field.resolve (/Users/ojongerius/repos/fcc-open-api/node_modules/graphql-tools/src/schemaGenerator.ts:780:18)
        at resolveFieldValueOrError (/Users/ojongerius/repos/fcc-open-api/node_modules/graphql/execution/execute.js:531:18)
        at resolveField (/Users/ojongerius/repos/fcc-open-api/node_modules/graphql/execution/execute.js:495:16)
        at /Users/ojongerius/repos/fcc-open-api/node_modules/graphql/execution/execute.js:364:18
      originalMessage: 'Expected a valid email, got "Ooops!"',
      message: 'Error in resolver Query.getUser\nExpected a valid email, got "Ooops!"' }

  console.error utils/asyncErrorHandler.js:6
    TypeError: Expected a valid email, got "Ooops!"
        at getUserPromise (/Users/ojongerius/repos/fcc-open-api/src/dataLayer/mongo/user.js:20:9)
        at new Promise (<anonymous>)
        at getUser (/Users/ojongerius/repos/fcc-open-api/src/dataLayer/mongo/user.js:17:10)
        at Object.<anonymous> (/Users/ojongerius/repos/fcc-open-api/node_modules/graphql-tools/src/schemaGenerator.ts:683:22)
        at class_1.<anonymous> (/Users/ojongerius/repos/fcc-open-api/node_modules/graphql-tools/src/schemaGenerator.ts:781:42)
        at step (/Users/ojongerius/repos/fcc-open-api/node_modules/graphql-tools/dist/schemaGenerator.js:51:23)
        at Object.next (/Users/ojongerius/repos/fcc-open-api/node_modules/graphql-tools/dist/schemaGenerator.js:32:53)
        at /Users/ojongerius/repos/fcc-open-api/node_modules/graphql-tools/dist/schemaGenerator.js:26:71
        at new Promise (<anonymous>)
        at Object.<anonymous>.__awaiter (/Users/ojongerius/repos/fcc-open-api/node_modules/graphql-tools/dist/schemaGenerator.js:22:12)
        at /Users/ojongerius/repos/fcc-open-api/node_modules/graphql-tools/src/schemaGenerator.ts:781:13
        at isAuthenticatedOnQuery (/Users/ojongerius/repos/fcc-open-api/src/graphql/resolvers/directives.js:24:32)
        at field.resolve (/Users/ojongerius/repos/fcc-open-api/node_modules/graphql-tools/src/schemaGenerator.ts:780:18)
        at resolveFieldValueOrError (/Users/ojongerius/repos/fcc-open-api/node_modules/graphql/execution/execute.js:531:18)
        at resolveField (/Users/ojongerius/repos/fcc-open-api/node_modules/graphql/execution/execute.js:495:16)
        at /Users/ojongerius/repos/fcc-open-api/node_modules/graphql/execution/execute.js:364:18

 PASS  graphql/resolvers/directive.test.js
  isAuthenticatedOnField
    ✓ should return null if authenication fails (3ms)
    ✓ should return the secretValue if authentication succeeds (1ms)
  isAuthenticatedOnQuery
    ✓ should throw an error is auth fails (1ms)
    ✓ should call next if auth succeeds

 PASS  dataLayer/mongo/user-dataLayer.test.js
  createUser
    ✓ should return a User object (24ms)
    ✓ should return user for accountLinkId if found in db (6ms)
  getUser
    ✓ should return a User object fo a valid request (3ms)
    ✓ should throw for a user not found (1ms)
    ✓ should throw if the supplied email is not valid (3ms)

Test Suites: 3 passed, 3 total
Tests:       18 passed, 18 total
Snapshots:   18 passed, 18 total
Time:        4.39s
Ran all test suites.
  console.error dataLayer/mongo/user.js:60
    Cannot create account, the unique id associated with this user is already in use a-very-unique-string-for-charlie@thebear.me

  console.error dataLayer/mongo/user.js:60
    Cannot create account, the unique id associated with this user is already in use a-very-unique-string-for-charlie@thebear.me
```

File: src/graphql/typeDefs/GDPR-User/mutation.js (hunk 20 above)
Comment (Bouncey) [reply src/graphql/typeDefs/GDPR-User/mutation.js:5]: 100% we should pull this off context. It was around this point I made a decision to get the PR ready for a merge and reduce scope creep. 

I'd happily remove it for now.

**Comment by Bouncey** [comment]
This is the reason for the `--silent` flag. 😁

**Comment by ojongerius** [comment]
😆 

Would it not be neater to expect and handle those errors? Could we use [tothrowerror](https://facebook.github.io/jest/docs/en/expect.html#tothrowerror) and or  [tothrowerrormatchingsnapshot](https://facebook.github.io/jest/docs/en/expect.html#tothrowerrormatchingsnapshot)?

File: test/integration/userFlow.test.js (hunk 17 above)
Comment (Bouncey) [reply test/integration/userFlow.test.js:152]: So, from my understanding, a snapshot is a text based representation of the thing being `expect()`'ed. It can be the output of a function, the state of a variable, anything!. `toMatchSnapshot()` effectively diffs the thing being `expect()`'ed against the stored snapshot to see if anything have changed. If things have changed, the test fails.

The most common use case is that instead of writing n * assertions, we take a dump of the thing being asserted instead and diff subsequent test output against it.

I guess there is a higher level of abstraction that comes with snapshots, you do not get a clear idea of the shape of the output unless you go and read the `.snap.js` files.

File: src/dataLayer/model/user.js (hunk 9 above)
Comment (raisedadead) [reply src/dataLayer/model/user.js]: Yes, that is correct. Each user from auth0 will have a unique ID that will be linked to a ObjectID on mongoDB for the User collection.

I am also inclined to name this `accountId` and use it as 'primary' key for everything.

File: src/dataLayer/mongo/user.js (hunk 7 above)
Comment (raisedadead) [reply src/dataLayer/mongo/user.js]: Haha! That's a good one.!

File: src/graphql/typeDefs/User/type.js [hunk 21]
```diff
@@ -1,7 +1,92 @@
...
-  email: String
+  email: String!
+  progressTimestamps: [ProgressTimestamp]!,
+  isBanned: Boolean
+  isCheater: Boolean
+  githubURL: String
```
Comment (raisedadead) [inline src/graphql/typeDefs/User/type.js]: `githubURL` => `github` ? like other socials?

File: src/dataLayer/model/user.js (hunk 9 above)
Comment (Bouncey) [reply src/dataLayer/model/user.js]: Yes it will, it will allow us to pull the right user record for the session

File: src/graphql/typeDefs/User/type.js (hunk 21 above)
Comment (Bouncey) [reply src/graphql/typeDefs/User/type.js]: Not required any more

File: package.json (hunk 8 above)
Comment (ojongerius) [reply package.json:30]: How can we have unexpected errors in our tests, if we wrote them? 🤔

File: test/integration/userFlow.test.js (hunk 17 above)
Comment (ojongerius) [reply test/integration/userFlow.test.js:152]: Ah interesting, thanks for the explanation. The n assertions would add value.

File: src/graphql/typeDefs/GDPR-User/mutation.js (hunk 20 above)
Comment (ojongerius) [reply src/graphql/typeDefs/GDPR-User/mutation.js:5]: 👍

**Comment by ojongerius** [comment]
I have no other concerns than that it looks messy to have uncaught errors in tests. Suppressing with a silent flag feels a little like a cover up.

**Comment by ojongerius** [comment]
Outstanding tasks AFAICS
- [ ] remove or improve deleteGDPRUser
- [ ] agree if tests should have uncaught errors

After that we should be good to merge, let's get this today or latest, tomorrow.

**Comment by raisedadead** [comment]
Yes, I agree.. I am going ahead and giving this a merge, to keep things moving! Thanks a lot @Bouncey this is some incredible work!

/cc @freeCodeCamp/open-api
